Improve sys.settrace to help support debugpy / pdb debugging

andrewleech opened the pull request:
Companion support for https://github.com/micropython/micropython-lib/pull/1022 & https://github.com/micropython/micropython-lib/pull/499

Requires micropython be built with `MICROPY_PY_SYS_SETTRACE` eg. [unix standard variant](https://github.com/micropython/micropython/blob/d7919ea71e7b7cc203ca984cc2f4a55019634835/ports/unix/variants/standard/mpconfigvariant.h#L32)

Includes:
* sys/settrace: Add sys._getframe() function.
  Refer to https://docs.python.org/3/library/sys.html#sys._getframe

* sys/settrace: Add frame.f_trace object support. 
  `f_trace` holds a python obj on the frame, used in `bdb` to track active trace function for call stack.

dpgeorge commented on py/modsys.c:
```diff
@@ -187,6 +187,21 @@ STATIC mp_obj_t mp_sys_settrace(mp_obj_t obj) {
+MP_DEFINE_CONST_FUN_OBJ_0(mp_sys_gettrace_obj, mp_sys_gettrace);
+
+// _getframe(): Return current frame object.
+STATIC mp_obj_t mp_sys_getframe(size_t n_args, const mp_obj_t *args) {
```
please call it `mp_sys__getframe`

dpgeorge commented on py/objfun.c:
```diff
@@ -359,6 +360,16 @@ void mp_obj_fun_bc_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+    if (attr == MP_QSTR___code__) {
+        mp_obj_fun_bc_t *self = MP_OBJ_TO_PTR(self_in);
+        mp_obj_code_t *code = MP_OBJ_TO_PTR(mp_obj_new_code(self->context, self->rc));
+        if (code != NULL) {
```
can code be null?

dpgeorge commented on py/profile.c:
```diff
@@ -212,17 +212,29 @@ STATIC void frame_print(const mp_print_t *print, mp_obj_t o_in, mp_print_kind_t
         case MP_QSTR_f_back:
             dest[0] = mp_const_none;
             if (o->code_state->prev_state) {
+                if (!o->code_state->prev_state->frame) {
```
did you find a case where this could be NULL?  maybe it needs a test...

dpgeorge commented on py/profile.c:
```diff
@@ -238,6 +250,12 @@ STATIC void frame_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+            dest[0] = o->trace_obj;
+            break;
+        case MP_QSTR_f_locals:
+            dest[0] = MP_OBJ_FROM_PTR(o->code->dict_locals);
```
is this needed?  locals in uPy doesn't work well

dpgeorge replied:
can we just return an empty dict instead?

dpgeorge commented on py/profile.c:
```diff
@@ -274,6 +292,7 @@ mp_obj_t mp_obj_new_frame(const mp_code_state_t *code_state) {
     o->lineno = mp_prof_bytecode_lineno(rc, o->lasti);
     o->trace_opcodes = false;
     o->callback = MP_OBJ_NULL;
+    o->trace_obj = MP_OBJ_NULL;
```
maybe call it `f_trace`

dpgeorge commented on py/profile.h:
```diff
@@ -60,7 +61,9 @@ mp_obj_t mp_obj_new_frame(const mp_code_state_t *code_state);
 
 // This is the implementation for the sys.settrace
 mp_obj_t mp_prof_settrace(mp_obj_t callback);
+mp_obj_t mp_prof_gettrace();
```
please put `void` in arg list

dpgeorge commented on py/profile.c:
```diff
@@ -313,6 +332,33 @@ mp_obj_t mp_prof_settrace(mp_obj_t callback) {
     return mp_const_none;
 }
 
+mp_obj_t mp_prof_gettrace() {
```
`void` in arg list

dpgeorge commented on py/profile.c:
```diff
@@ -313,6 +332,33 @@ mp_obj_t mp_prof_settrace(mp_obj_t callback) {
+    return prof_trace_cb;
+}
+
+mp_obj_t mp_prof_get_frame(int depth) {
```
can we change `int` to `size_t` to indicate it can't be negative?

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.

pi-anl committed:
sys/settrace: Improve settrace to support debugpy/pdb debugging.

Add sys.gettrace() to mirror settrace, add sys._getframe() for stack
frame inspection, and enable sys.settrace() by default on unix
standard variant. These additions provide the baseline infrastructure
needed for pdb and debugpy to function on MicroPython.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>
